### PR TITLE
Fix missing log_gamma in nfa()

### PR DESCRIPTION
**Pull request by ibaiGorordo**
Comparing the [nfa function](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/lsd.cpp#L1074) with the function in the [binomial_nfa repository](https://github.com/rafael-grompone-von-gioi/binomial_nfa/blob/main/C99/log_binomial_nfa.c#L152), the first log_gamma call is missing.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by alalek** [comment]
@ibaiGorordo Thank you for the contribution!

 Need to check and update related tests.

**Comment by ibaiGorordo** [comment]
Hi, 

I saw that the test failed always when testing with the rotatedRect: https://github.com/opencv/opencv/blob/6a7d54f55074feb212a04a321663b1a897f3ad57/modules/imgproc/test/test_lsd.cpp#L156.

I did some testing to try to understand why it was failing, but then I tested the same test with the original LSD method it was also failing the test. The reason why the incorrect method is working is that doing the refinement with LSD_REFINE_ADV returns almost the same result as using LSD_REFINE_STD because rect_improve is basically almost doing nothing.

The issue is that since the log_gamma is missing the value of log1term() becomes a large negative value and therefore term becomes 0.0 in many cases if not always.
https://github.com/opencv/opencv/blob/6a7d54f55074feb212a04a321663b1a897f3ad57/modules/imgproc/src/lsd.cpp#L1077

Here is the calculations with the current code for one of the lines of the treated rect test:
`n+1: 815.000000, log_gamma(n+1): 4645.665759, log1term: -3832.690976, term: 0.000000`

And this are the same calculations with the corrected code:
`n+1: 815.000000, log_gamma(n+1): 4645.665759, log1term: -2.025217, term: 0.131965`

As a result, with the current code, it in many cases (if not always), this if statement is true, and therefore the value of nfa in the current code is either `-log1term / M_LN10 - LOG_NT` or `-LOG_NT` depending on `k > n * p`.

https://github.com/opencv/opencv/blob/6a7d54f55074feb212a04a321663b1a897f3ad57/modules/imgproc/src/lsd.cpp#L1079-L1083

In summary, the ADV version of the rotatedRect test is not correct since the original LSD method does not detect the lines of the rotatedRect in all the epochs when doing the rect_improve function.

**Comment by ibaiGorordo** [comment]
I have found the other issue (there might be more). The rect_nfa function was not checking the pixels inside the rectangle correctly. With the last changes to the rect_nfa function, now it checks the same pixels as the original paper. In the image below I fill a rotated rectangle using the original paper method (blue), the old OpenCV method (green), and the fixed code (red). As you can see, the old calculation had some logic issues.
rect_fill_comparison

I have tested the fix with the rotatedRect test example, and now it detects all the lines properly (compared to before when sometimes some of the lines were missing).

**Comment by ibaiGorordo** [comment]
Closing PR and opening a new one due to a mistake merging opencv:4.x
